File: haskell-playground/src/Datastructures/Trees/TreeSize.hs
module Datastructures.Trees.TreeSize (treeSize) where

import Datastructures.Trees.MyBinaryTree (Tree (..))

-- From Wikipedia:
-- Size of a tree = Number of nodes in the tree.

treeSize :: Num p => Tree a -> p
treeSize EmptyTree = 0
treeSize (Node a EmptyTree EmptyTree ) = 1
treeSize tree@(Node a left right) = 1 + treeSize left + treeSize right

File: go-playground/02-go-by-examples/05mydatastructures/trees/tree_size.go
package mydatastructures

func TreeSize(bst BinarySearchTree) int {
	if bst == (BinarySearchTree{}) {
		return 0
	}
	if bst.Left == nil && bst.Right == nil {
		return 1
	}
	if bst.Left == nil && bst.Right != nil {
		return 1 + TreeSize(*bst.Right)
	}
	if bst.Left != nil && bst.Right == nil {
		return 1 + TreeSize(*bst.Left)
	}
	return 1 + TreeSize(*bst.Left) + TreeSize(*bst.Right)
}
